contracts/state-validator/src/verifications/finalize_withdrawal.rs

```diff
@@ -0,0 +1,111 @@
+use core::result::Result;
+
+use gw_state::ckb_smt::smt::{Pair, Tree};
+use gw_utils::{
+    ckb_std::debug,
+    error::Error,
+    gw_common::H256,
+    gw_types::{
+        bytes::Bytes,
+        packed::{GlobalState, RawL2Block, RollupConfig, RollupFinalizeWithdrawalReader},
+        prelude::{Builder, Entity, Unpack},
+    },
+};
+
+pub mod last_finalized_withdrawal;
+mod user_withdrawal_cells;
+
+#[must_use]
+pub fn verify(
+    rollup_type_hash: &H256,
+    config: &RollupConfig,
+    args: RollupFinalizeWithdrawalReader,
+    prev_global_state: GlobalState,
+    post_global_state: GlobalState,
+) -> Result<(), Error> {
+    debug!("verify finalized withdrawal");
+
+    // Check global state version is 2
+    if 2 != prev_global_state.version_u8() || 2 != post_global_state.version_u8() {
+        debug!("global state invalid version");
+        return Err(Error::InvalidGlobalStateVersion);
+    }
+
+    // Check global state last_finalized_block_number
+    let expected_last_finalized_block_number = {
+        let block_number = prev_global_state.block().count().unpack().saturating_sub(1);
+        block_number.saturating_sub(config.finality_blocks().unpack())
+    };
+    debug!(
+        "expected last finalize block number {}",
+        expected_last_finalized_block_number
+    );
+    let last_finalized_block_number = prev_global_state.last_finalized_block_number().unpack();
+    if expected_last_finalized_block_number != last_finalized_block_number {
+        debug!("global state wrong last finalized block number");
+        debug!("finalized block number {}", last_finalized_block_number);
+        return Err(Error::InvalidPostGlobalState);
+    }
```

Comment:
> We should verify post_global_state but not prev_global_state.

`RollupFinalizeWithdrawal` action doesn't submit new block, `last_finalized_block_number` should base on `prev_global_state` not `post_global_state`.

> However, I believe that the verification of last_finalized_block_number is redundant since it has been covered by the verification of [expected_post_global_state](https://github.com/nervosnetwork/godwoken-scripts/pull/126/files#diff-0335e07e97d95c2e077fcc07105c5d482717f4f2f7589999f680b12cde057e2fR70-R77) below.

They are different check. 

Check `last_finalized_block_number` in `prev_global_state` is just in case that `last_finalized_block_number` is valid. 
`expected_post_global_state` is used to verify that `RollupFinalizeWithdrawal` only change `last_finalized_withdrawal` field.

IMO, double check is ok here. It's contract code, you can't be too careful.

Reply:
But `prev_global_state` has already been committed on L1, what is the purpose of verifying `prev_global_state`?

Reply:
But you can submitted different ```post_global_state```, here we want to make sure `last_finalized_block_number` is right.

Reply:
Make sure `post_global_state`'s `last_finalized_block_number` is right by comparing `prev_global_state.last_finalized_block_number()` and `prev_global_state.block().count() - config.finality_blocks()`?